nnvm/src/top/tensor/reduce.cc

```diff
@@ -170,12 +170,19 @@ Example::
   "FGradient", [](const NodePtr& n,
                   const std::vector<NodeEntry>& ograds){
     const ReduceParam& param = nnvm::get<ReduceParam>(n->attrs.parsed);
-    std::ostringstream axis; axis << param.axis;
+    bool exclude = param.exclude;
+    TShape p_axis = param.axis;
+    if (!param.exclude && param.axis.ndim() == 0) {
+      exclude = true;
+      p_axis = TShape();
+    }
+    std::ostringstream axis; axis << p_axis;
     return std::vector<NodeEntry>{
       MakeNode("expand_like", n->attrs.name + "_grad",
                {ograds[0], n->inputs[0]},
                {{"axis", axis.str()},
-                {"exclude", std::to_string(param.exclude)}})
+                {"keepdims", std::to_string(param.keepdims)},
```

Comment:
Passing `keepdims` to expand_like looks a bit strange since expand_like is not a reduce operation actually.  Whether the parameter is specified or not, the output dimension does not changed.

An alternative I came up with was squeezing the output of sum beforehand like as follows.  This might be less efficient, but looks easier to understand what we want to do.
```
  if (param.keepdims) {
    NodeEntry squeezed = MakeNode("squeeze", n->attrs.name + "_grad_sqz",
                                  {ograds[0]},
                                  {{"axis", axis.str()}});
    return std::vector<NodeEntry>{
      MakeNode("expand_like", n->attrs.name + "_grad",
               {squeezed, n->inputs[0]},
               {{"axis", axis.str()},
                {"exclude", std::to_string(exclude)}})};
  }
```

Reply:
> Well, expand_like is not a reduction operation, but it might be considered dual to reduction: when given the same parameters as a reduction operation, it expands exactly the same dimensions as the reduction operation reduces.

I'd agree that those are dual, but I think it's not necessary to have the `keepdims` parameter in expand operators.  I'll send a comment about it later.

> Your solution needs adding an exclude parameter to the squeeze operation, otherwise it won't work when exclude=True. 

Having the exclude option in squeeze sounds good to me.  But if you don't prefer it, I'm fine with broadcasting in expand_like. :)